### PR TITLE
TASK: Lock events table before committing

### PR DESCRIPTION
**Pull request by stephanschuler**
There are situations where two processes try to write data to the event
store simultaneously.

sequence number | event | source
----------- | --------------- | -------
1 |  do first thing        |first process
2 |do second thing       |first process
3 | do something different|second process
4  |do third thing        |first process
5 | do fourth thing       |first process

The second process starts second but will finish first.

So there is a point in time the event store looks like this:

sequence number | event | source
----------- | --------------- | -------
3 | do something different | second process

If now a listener starts catching up, it will be remered as
highest applied sequence number = 3.

As soon as the first process finishes, another catch up happens starting
at sequence number 4.
So in this very situation, sequence numbers 1 and 2 are recorded and
even recorded in the right order but never played.

### CODE REVIEW

File: Classes/EventStore/Storage/Doctrine/DoctrineEventStorage.php [hunk 1]
```diff
@@ -147,6 +147,11 @@ public function commit(StreamName $streamName, WritableEvents $events, int $expe
         while (true) {
             $this->reconnectDatabaseConnection();
             $this->connection->beginTransaction();
+
+            # `SELECT FOR UPDATE` write-locks affected data, `SELECT MAX() FOR UPDATE` write-locks the table.
+            # This should be "LOCK TABLE" but "LOCK TABLE" doesn't play nice with transactions.
```
Comment (paxuclus) [inline Classes/EventStore/Storage/Doctrine/DoctrineEventStorage.php]:
```suggestion
            # This should be "LOCK TABLE" but "LOCK TABLE" doesn't play nicely with transactions.
```

File: Classes/EventStore/Storage/Doctrine/DoctrineEventStorage.php [hunk 2]
```diff
@@ -147,6 +147,11 @@ public function commit(StreamName $streamName, WritableEvents $events, int $expe
...
+            $this->connection->query('SELECT MAX(sequencenumber) FROM ' . $this->eventTableName . ' FOR UPDATE')->fetchAll();
```
Comment (paxuclus) [inline Classes/EventStore/Storage/Doctrine/DoctrineEventStorage.php:153]:
```suggestion
            $this->connection->executeQuery('SELECT MAX(sequencenumber) FROM ' . $this->eventTableName . ' FOR UPDATE');
```
Do we really need the fetchAll()?

Comment (paxuclus) [reply Classes/EventStore/Storage/Doctrine/DoctrineEventStorage.php:153]: Oh I see, `SELECT` yields results, so it should be *fetch*

**Review by paxuclus — APPROVED** [review]

File: Classes/EventStore/Storage/Doctrine/DoctrineEventStorage.php (hunk 1 above)
Comment (stephanschuler) [reply Classes/EventStore/Storage/Doctrine/DoctrineEventStorage.php]: Done.

**Pull request merged**